lldb/source/Target/Target.cpp

```diff
@@ -1538,73 +1538,35 @@ void Target::SetExecutableModule(ModuleSP &executable_sp,
-        if (image_module_sp) {
-          added_modules.AppendIfNeeded(image_module_sp, false);
-          ObjectFile *objfile = image_module_sp->GetObjectFile();
-          if (objfile) {
-            // Create a local copy of the dependent file list so we don't have
-            // to lock for the whole duration of GetDependentModules.
-            FileSpecList dependent_files_copy;
-            {
-              std::lock_guard<std::mutex> guard(dependent_files_mutex);
-              dependent_files_copy = dependent_files;
-            }
-
-            // Remember the size of the local copy so we can append only the
-            // modules that have been added by GetDependentModules.
-            const size_t previous_dependent_files =
-                dependent_files_copy.GetSize();
-
-            objfile->GetDependentModules(dependent_files_copy);
-
-            {
-              std::lock_guard<std::mutex> guard(dependent_files_mutex);
-              for (size_t i = previous_dependent_files;
-                   i < dependent_files_copy.GetSize(); ++i)
-                dependent_files.AppendIfUnique(
-                    dependent_files_copy.GetFileSpecAtIndex(i));
-            }
-          }
-        }
-      };
-
-      executable_objfile->GetDependentModules(dependent_files);
+    if (!executable_objfile || !load_dependents)
+      return;
 
-      llvm::ThreadPoolTaskGroup task_group(Debugger::GetThreadPool());
-      for (uint32_t i = 0; i < dependent_files.GetSize(); i++) {
-        // Process all currently known dependencies in parallel in the innermost
-        // loop. This may create newly discovered dependencies to be appended to
-        // dependent_files. We'll deal with these files during the next
-        // iteration of the outermost loop.
-        {
-          std::lock_guard<std::mutex> guard(dependent_files_mutex);
-          for (; i < dependent_files.GetSize(); i++)
-            task_group.async(GetDependentModules,
-                             dependent_files.GetFileSpecAtIndex(i));
-        }
-        task_group.wait();
+    llvm::ThreadPoolTaskGroup task_group(Debugger::GetThreadPool());
+    ModuleList added_modules;
+    FileSpecList dependent_files;
+    executable_objfile->GetDependentModules(dependent_files);
+    
+    for (uint32_t i = 0; i < dependent_files.GetSize(); i++) {
+      auto dependent_file_spec = dependent_files.GetFileSpecAtIndex(i);
+      FileSpec platform_dependent_file_spec;
+      if (m_platform_sp)
+        m_platform_sp->GetFileWithUUID(dependent_file_spec, nullptr,
+                                       platform_dependent_file_spec);
+      else
+        platform_dependent_file_spec = dependent_file_spec;
+
+      ModuleSpec module_spec(platform_dependent_file_spec, m_arch.GetSpec());
+      auto image_module_sp = GetOrCreateModule(
+          module_spec, false /* notify */,
+          [&](auto preloadSymbols) { task_group.async(preloadSymbols); });
```

Comment:
I think it would be good to gate the parallel loading of modules here based on the with the `target.parallel-module-load` setting.